### ISSUE TITLE
[GRPO] per_token_logps.detach() should come from the old_policy – ratio is always 1. Is this intended?

### ISSUE THREAD

**Issue by SeungyounShin** [huggingface/trl#2769]
I noticed in [[this code](https://github.com/huggingface/trl/blob/af4ad47035529164799be10f3fe558ee642a9880/trl/trainer/grpo_trainer.py#L567)](https://github.com/huggingface/trl/blob/af4ad47035529164799be10f3fe558ee642a9880/trl/trainer/grpo_trainer.py#L567) the following snippet:

```python
# x - x.detach() allows for preserving gradients from x
advantages = inputs["advantages"]
per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
per_token_loss = -(per_token_loss - self.beta * per_token_kl)
loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()
```

Shouldn't `per_token_logps.detach()` be computed from the old policy? As written, the ratio becomes always 1. Is this the intended behavior, or am I missing something?

**Comment by SeungyounShin** [comment]
Here is how OpenRLHF dealing with ratio in PPO

https://github.com/OpenRLHF/OpenRLHF/blob/631b0bcb7c14ec2fd4117a43a661903ed60d26cc/openrlhf/trainer/ppo_trainer.py#L366

**Comment by SeungyounShin** [comment]
This is related to https://github.com/huggingface/trl/issues/2608


I fully grasped why ratio always 1
